File: src/main/java/org/distril/beengine/material/block/BlockRegistry.java
package org.distril.beengine.material.block;

import org.distril.beengine.material.Material;
import org.distril.beengine.material.block.impl.NoopBlock;

import java.util.EnumMap;
import java.util.Map;

public class BlockRegistry {

	private final Map<Material, Block> byMaterial = new EnumMap<>(Material.class);

	public void init() {
		// todo
	}

	public void register(Material material, Block block) {
		this.byMaterial.put(material, block);
	}

	public Block getBlockFromState(BlockState state) {
		var states = state.getStates();

		var identifier = states.getString("name");

		var material = Material.fromIdentifier(identifier);
		return this.from(material, state);
	}

	@SuppressWarnings("unchecked")
	public <T extends Block> T from(Material material) {
		return (T) this.byMaterial.getOrDefault(material, new NoopBlock(material)).clone();
	}

	@SuppressWarnings("unchecked")
	public <T extends Block> T from(Material material, BlockState state) {
		var from = this.from(material);
		from.setState(state);
		return (T) from;
	}
}

File: src/main/kotlin/org/distril/beengine/material/block/BlockRegistry.kt
package org.distril.beengine.material.block

import org.distril.beengine.material.Material
import org.distril.beengine.material.block.impl.NoopBlock
import java.util.*

class BlockRegistry {

	private val byMaterial: MutableMap<Material, Block> = EnumMap(Material::class.java)

	fun init() {
		// todo
	}

	fun register(material: Material, block: Block) {
		if (material.isBlock) this.byMaterial[material] = block
	}

	fun getBlockFromState(state: BlockState): Block {
		val identifier = state.states.getString("name")
		val material = Material.fromIdentifier(identifier)!!
		return this.from(material, state)
	}

	@Suppress("UNCHECKED_CAST")
	fun <T : Block> from(material: Material, state: BlockState? = null): T {
		return (this.byMaterial.getOrDefault(material, NoopBlock(material)).clone() as T).apply {
			if (state != null) this.state = state
		}
	}
}
